rpc/codec/json.go

```diff
@@ -2,71 +2,130 @@ package codec
 
 import (
 	"encoding/json"
+	"fmt"
 	"net"
+	"time"
 
 	"github.com/ethereum/go-ethereum/rpc/shared"
 )
 
 const (
-	MAX_RESPONSE_SIZE = 64 * 1024
+	READ_TIMEOUT      = 15 // read timeout in seconds
+	MAX_REQUEST_SIZE  = 1024 * 1024
+	MAX_RESPONSE_SIZE = 1024 * 1024
 )
 
 // Json serialization support
 type JsonCodec struct {
 	c net.Conn
-	d *json.Decoder
-	e *json.Encoder
 }
 
 // Create new JSON coder instance
 func NewJsonCoder(conn net.Conn) ApiCoder {
 	return &JsonCodec{
 		c: conn,
-		d: json.NewDecoder(conn),
-		e: json.NewEncoder(conn),
 	}
 }
 
 // Serialize obj to JSON and write it to conn
-func (self *JsonCodec) ReadRequest() (*shared.Request, error) {
-	req := shared.Request{}
-	err := self.d.Decode(&req)
-	if err == nil {
-		return &req, nil
+func (self *JsonCodec) ReadRequest() (requests []*shared.Request, isBatch bool, err error) {
+	bytesInBuffer := 0
+	buf := make([]byte, MAX_REQUEST_SIZE)
+
+	deadline := time.Now().Add(READ_TIMEOUT * time.Second)
+	if err := self.c.SetDeadline(deadline); err != nil {
+		return nil, false, err
+	}
+
+	for {
+		n, err := self.c.Read(buf[bytesInBuffer:])
+		if err != nil {
+			self.c.Close()
+			return nil, false, err
+		}
+
+		bytesInBuffer += n
+
+		singleRequest := shared.Request{}
+		err = json.Unmarshal(buf[:bytesInBuffer], &singleRequest)
+		if err == nil {
+			requests := make([]*shared.Request, 1)
+			requests[0] = &singleRequest
+			return requests, false, nil
+		}
+
+		requests = make([]*shared.Request, 0)
+		err = json.Unmarshal(buf[:bytesInBuffer], &requests)
+		if err == nil {
+			return requests, true, nil
+		}
```

Comment:
You can avoid the double Unmarshal by scanning for the first JSON character before decoding.
Sketch:

```go
buf := bufio.NewReader(conn)
var val interface{}
loop:
for {
    b, err := buf.ReadByte()
    if err != nil {
        return err
    }
    switch b {
    case ' ', '\n', '\t':
        // skip whitespace before value
        continue loop
    default:
         if b == '[' {
              val = new([]*shared.Request)
         } else {
              val = new(shared.Request)
         }
         buf.UnreadByte()
         break loop
     }
}
err := json.NewDecoder(buf).Decode(val)
return val, err